client/index.ts

```diff
@@ -167,14 +167,19 @@ class OAuth2Client {
 
   async verifyJWT<T = unknown>(accessToken: string, algo: string): Promise<T> {
     await this.setOpenIDConfiguration();
-    await this.setJWKS();
 
-    return new Promise((resolve, reject) => {
-      const [header, payload] = accessToken.split(".");
+    const [header, payload] = accessToken.split(".");
 
-      const { alg, kid } = decodeJWTPart<{ alg: string; kid: string }>(header);
-      const { aud } = decodeJWTPart<{ aud: string }>(payload);
+    const { alg, kid } = decodeJWTPart<{ alg: string; kid: string }>(header);
+    const { aud } = decodeJWTPart<{ aud: string }>(payload);
+
+    if (kid) {
+      if (!this.jwks) {
+        await this.setJWKS();
```

Comment:
Considering the other time we do something like this for `this.setOpenIdConfiguration()` this feels a bit weird because the two methods are not doing it the same way.

But I think we can do better for both of them.

At the moment we have:

method 1: always call a setter (`this.setOpenIDConfiguration`) and return an empty promise if it's already in cache then use a private attribute to get the data.
method 2: call a setter (`this.setJWKS()`) in advance of when we really need it but only if a private key has not been set.

I think it's better to use only local variables and rely less as possible on private attributes of the class (this is debattable) so here what I propose as a third method: get rid of setters and only have getters that will do the fetch if needed.

Or in code:

```ts
const openIDConfiguration = this.getOpenIDConfiguration();

if (kid) {
  const jwks = this.getJWKS();
  // ...
```

```ts
getOpenIDConfiguration() {
  if (this.openIDConfiguration) {
    return Promise.resolve(this.openIDConfiguration);
  }
  return fetch().then(result => {
    this.openIDConfiguration = result;
    return result;
  })
}
```

What do you think?

Reply:
I'm confused. 
```ts
const openIDConfiguration = this.getOpenIDConfiguration();

if (kid) {
  const jwks = this.getJWKS();
  // ...
```

This means we won't store it, right?

Reply:
```ts
getJWKS() {
  if (this.jwks) {
    return Promise.resolve(this.jwks);
  }
}
```

Reply:
And we fetch it if this condition is false. Got it.
I like it.

Reply:
This is like having a `getter` and a `setter` at the same time.

Reply:
Sounds good to me too. I'm on it 🚀

Reply:
More like having a getter with its own cache 😅 but semantics pedantics